### ISSUE TITLE
Authenticating with a GitHub App not working with custom URL

### ISSUE THREAD

**Issue by AlexHff** [PyGithub/PyGithub#2447]
Hello there,

It seems like the `base_url` passed to the default constructor is not used to fetch the access token for app authentication.

Here's an example code:
```
app_id = "XXX"
app_private_key = "XXX"
installation_id = "XXX"
g = github.Github(
    base_url="https://custom-github-url.com"
    app_auth=github.AppAuthentication(
        app_id=app_id,
        private_key=app_private_key,
        installation_id=installation_id,
    ),
)
```

Whatever `base_url` is passed, the following error occurs:
`requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: XXX`

As can be seen in the stack trace, the host is the one defined [here](https://github.com/PyGithub/PyGithub/blob/5e27c10a3140c3b9bbf71a0b71c96e71e1e3496c/github/Consts.py#L135), which is `DEFAULT_BASE_URL = "https://api.github.com"`.

I suspect the error comes from [here](https://github.com/PyGithub/PyGithub/blob/5e27c10a3140c3b9bbf71a0b71c96e71e1e3496c/github/Requester.py#L377). The param `base_url` is not passed to `GithubIntegration` which therefore uses the default url.

**Comment by EnricoMi** [comment]
This is been reported in #2420 and fixed in #2425. Will be released in [v0.58.1](https://github.com/PyGithub/PyGithub/milestone/39).

**Comment by goldstar611** [comment]
I tested with master today and while `base_url` is propagated, `verified` is not. `verified` is a bool or string and the string is needed when the enterprise installation uses a SSL/TLS certificate that is not issued by a root ca in pythons included ca bundle.

We're getting there though!

**Comment by EnricoMi** [comment]
The fix has been released.

**Comment by EnricoMi** [comment]
> but then while refreshing the access token, a check for expiring access token is done which then refreshes the access token, which checks the token ... until recursion limit is met.

@goldstar611 I could not reproduce this issue, and looking into the code, I could not see how this would recurse. The refresh request creates its own new requester instance. Can you provide a stack trace? That would help to investigate the issue.

**Comment by goldstar611** [comment]
I tested both the `Github` class with a personal access token as well as the `GithubIntegration` class with a private key and did not see this issue again.

I remember what happened however. I changed the host name (sometimes we use a domain suffix inside out LAN) and that caused the Github appliance to send a redirect. This can be reproduced with the following test case but, IMO, it's user error.

```python
from github import Github

# Github Enterprise with custom hostname
g = Github(base_url="https://www.github.com/api/v3",
#                            ^^^^ notice the www.
           login_or_token="doesn't matter what it is",)
print(g.get_user().name)
```

**Comment by EnricoMi** [comment]
Thank you for the input. I think PyGithub should identify this situation and raise an error and not follow the redirection. This is going to be fixed in #2524.
